smolyay/samples.py

```diff
@@ -508,3 +511,583 @@ def _create(self):
+
+    Parameters
+    ----------
+    point_sets : list of :class:UnidimensionalPointSet
+        unique set of 1D points
+    """
+
+    def __init__(self, point_sets):
+        super().__init__()
+        self._point_sets = point_sets
+        self._indexes = []
+
+    @property
+    def domain(self):
+        """numpy.ndarray: domain of the point set."""
+        return numpy.array([up.domain for up in self.point_sets])
+
+    @domain.setter
+    def domain(self, value):
+        domain = numpy.sort(numpy.array(value, ndmin=2), axis=1)
+        if domain.ndim != 2 or domain.shape[1] != 2:
+            raise TypeError("Domain must have size (num_dimensions, 2)")
+        if any(domain[:, 0] >= domain[:, 1]):
+            raise ValueError("Lower bound must be less than upper bound")
+        if len(self.point_sets) != domain.shape[0]:
+            raise IndexError("Domain does not match number of point sets")
+        if not numpy.array_equal(self._domain, domain):
+            self._domain = domain
+            for ps, d in zip(self.point_sets, domain):
+                ps.domain = d
+            self._valid_cache = False
+
+    @property
+    def point_sets(self):
+        """:class:UnidimensionalPointSet: set of unique points"""
+        return self._point_sets
+
+
+class TensorProductPointSet(PointSetProduct):
+    """Generate a grid using all combinations of unidimensional point sets
+
+    Depending on the dimensionality, points and provided by
+    :class:`UnidimensionalPointSet`, make full tensor grids.
+    :meth:`generates_points` generates a full tensor grid
+
+    Parameters
+    ----------
+    unique_points : list of :class:UnidimensionalPointSet
+        unique set of 1D points
+    """
+
+    def _create(self):
+        """Generating the tensor grid
+
+        Using the parameters of the class, generate a grid set using all
+        combinations of unidimensional points
+        """
+        num_points = numpy.prod([len(p) for p in self._point_sets])
+        self._points = numpy.zeros((num_points, len(self._point_sets)), dtype=float)
+        for i, point in enumerate(itertools.product(*self._point_sets)):
+            self._points[i] = point
+
+class SmolyakSparseProductPointSet(PointSetProduct):
+    """Generate a grid using sparse combinations of unidimensional point sets
+
+    Depending on the dimensionality, points and provided by
+    :class:`UnidimensionalPointSet`, make full tensor grids.
+    :meth:`generates_points` generates a full tensor grid
+
+    Parameters
+    ----------
+    unique_points : list of :class:UnidimensionalPointSet
+        unique set of 1D points
+    """
+
+    def _create(self):
+        """Generating the tensor grid
+
+        Using the parameters of the class, generate a grid set using all
+        combinations of unidimensional points
+        """
+        # reset points
+        self._points = None
+        # find number of levels per dimension
```

Comment:
Please add some whitespace to this function to make it easier to read. Probably a line before each of your comments would make sense.
```suggestion

        # find number of levels per dimension
```